### PR TITLE
Resolving Remember username issue

### PR DESCRIPTION
**Pull request by ajayavado**
Resolving Remember username issue.

Steps to replicate:: 
1) Log in with admin account by clicking remeberusername check box
2) Search another user
3) Use option login as and log as that user
4) Logout

Expected behaviour:
When we log out of the user profile, It should go back to the log in screen and have admin username displayed ready to log back in with my admin account.

Note: Patch file is available within JIRA ticket.

### CODE REVIEW

**Review by MrJons — COMMENTED** [review]
Do we need to only create this cookie for someone with 'manager' privileges? I think if we do it across the board we might be overriding any cookie security that's currently in place for regular users?

**Comment by ajayavado** [comment]
Hi @MrJons 
We checked the code for exiting cookie creation (MOODLEID1), and have replicated the same with a new cookie.

**Comment by MrJons** [comment]
Yes but that will be deleting cookies for certain circumstances, what we are doing is working around that behaviour to continue to provide data. Given we don't know all the circumstances in which the original cookie is being deleted it doesn't seem wise to be implementing this work around for all users.

For instance what if someone is on a public computer, log out thinking everything is good but we have secretly persisted their username?

File: classes/helper/login_helper.php [hunk 1]
```diff
@@ -519,16 +549,25 @@ public function is_username_set_in_auth_global_vars() {
...
-    public function clear_username_in_auth_global_vars() {
+    public function clear_username_in_auth_global_vars()
+    {
         global $frm;
         $frm = null;
         set_moodle_cookie('');
+
+        $cookiename = 'REMEMBERUNAME';
+        if (isset($_COOKIE[$cookiename])) {
+            // Delete old cookie.
```
Comment (udayavado) [inline classes/helper/login_helper.php]: Inline comment?

File: index.php [hunk 2]
```diff
@@ -94,10 +94,34 @@
...
         $values = $passwordform->get_data();
         $helper->set_passform_params_in_auth_global_vars($values);
+
+        //Setting loginas cookie
+        if (!empty($values->rememberme)) {
+            global $CFG;
+            // Set username cookie for 60 days.
+            $cookiesecure = is_moodle_cookie_secure();
+            $Loggedinusername = $values->username;
+            setcookie('REMEMBERUNAME', rc4encrypt($Loggedinusername), time() + (DAYSECS * 60), $CFG->sessioncookiepath,
```
Comment (udayavado) [inline index.php]: Not compulsory, just to make it more readable can we break the args in multiple lines

File: index.php [hunk 3]
```diff
@@ -94,10 +94,34 @@
...
 
+    global $CFG;
+    if (!empty($CFG->rememberusername)) {
+        $cookiename = 'REMEMBERUNAME';
+        if (isset($_COOKIE[$cookiename])) {
+            $username = rc4decrypt($_COOKIE[$cookiename]);
+        }
+        global $frm;
+        $frm->username = $username;
+        $passwordform->set_data(array(
```
Comment (udayavado) [inline index.php:106]: use [] instaed of array() ?

**Review by udayavado — APPROVED** [review]

File: classes/helper/login_helper.php [hunk 4]
```diff
@@ -519,16 +549,25 @@ public function is_username_set_in_auth_global_vars() {
...
         return isloggedin() and !isguestuser();
     }
 
     /**
      * Clear the username from
      */
-    public function clear_username_in_auth_global_vars() {
+    public function clear_username_in_auth_global_vars()
+    {
         global $frm;
```
Comment (vijaysavakhande) [inline classes/helper/login_helper.php:530]: No need to declared global again once declared in the script, unless within function call.

File: index.php [hunk 5]
```diff
@@ -94,10 +94,34 @@
         ));
     }
 
+    global $CFG;
+    if (!empty($CFG->rememberusername)) {
+        $cookiename = 'REMEMBERUNAME';
+        if (isset($_COOKIE[$cookiename])) {
+            $username = rc4decrypt($_COOKIE[$cookiename]);
+        }
+        global $frm;
```
Comment (vijaysavakhande) [inline index.php:104]: re-declared global, not necessary

**Review by vijaysavakhande — APPROVED** [review]
works fine at locally, just need little bit cleanup.

File: index.php [hunk 6]
```diff
@@ -94,10 +94,34 @@
         ));
     }
 
+    global $CFG;
+    if (!empty($CFG->rememberusername)) {
+        $cookiename = 'REMEMBERUNAME';
```
Comment (somnathavado) [inline index.php]: can we define a constant for 'REMEMBERUNAME' cookie, as we used this on multiple places in same file.
also if block can be written as 
define constant for 
const  COOKIE_REMEMBERUNAME = 'REMEMBERUNAME'; // define class const or use define function for it
if (!empty($CFG->rememberusername) && isset($_COOKIE[self::COOKIE_REMEMBERUNAME]) ) {
        global $frm;
        $frm->username =  rc4decrypt($_COOKIE[self::COOKIE_REMEMBERUNAME]);;
        $passwordform->set_data( [ 'username' => $frm->username, 'rememberme' => 1 ] );
    }

File: index.php [hunk 7]
```diff
@@ -94,10 +94,34 @@
...
         $passwordform->is_validated()) {
         $values = $passwordform->get_data();
         $helper->set_passform_params_in_auth_global_vars($values);
+
+        //Setting loginas cookie
+        if (!empty($values->rememberme)) {
+            global $CFG;
+            // Set username cookie for 60 days.
+            $cookiesecure = is_moodle_cookie_secure();
+            $Loggedinusername = $values->username;
```
Comment (somnathavado) [inline index.php]: can we directly pass is_moodle_cookie_secure(), $values->username values for this in setcookie function.
we don't need local variable for it.

**Review by somnathavado — CHANGES_REQUESTED** [review]

File: classes/helper/login_helper.php (hunk 1 above)
Comment (ajayavado) [reply classes/helper/login_helper.php]: Removed inline comment.

File: index.php (hunk 6 above)
Comment (ajayavado) [reply index.php]: Done, Please review.

File: index.php (hunk 7 above)
Comment (ajayavado) [reply index.php]: Done, Please review.

**Review by somnathavado — APPROVED** [review]

**Comment by ajayavado** [comment]
> Yes but that will be deleting cookies for certain circumstances, what we are doing is working around that behaviour to continue to provide data. Given we don't know all the circumstances in which the original cookie is being deleted it doesn't seem wise to be implementing this work around for all users.
> 
> For instance what if someone is on a public computer, log out thinking everything is good but we have secretly persisted their username?

Hi @MrJons 
Have implemented manager , capability check.
Please review.

**Comment by qamayank** [comment]
Tested at locally and working fine for me.

**Review by qamayank — APPROVED** [review]
Tested at locally and working fine for me.

**Comment by MrJons** [comment]
Im getting these Warnings...
image

Otherwise looks to work

**Comment by ajayavado** [comment]
> Im getting these Warnings...
> image
> 
> Otherwise looks to work

Hi @MrJons 
Can you please pull code, and run purgecache file.
Its language string not getting loaded.
Please check.

**Review by MrJons — APPROVED** [review]

**Review by LoneBeta — APPROVED** [review]

File: index.php [hunk 8]
```diff
@@ -102,6 +117,16 @@
 }
 
 if ($helper->authenticate()) {
+    if ($helper->is_user_already_loggedin()) {
+        //Setting loginas cookie
+        $context = context_system::instance();
+        $genratecookie = has_capability('local/signin:manage', $context);
```
Comment (LoneBeta) [inline index.php:123]: Can we change this to the correct spelling, should be "generatecookie".

Comment (ajayavado) [reply index.php:123]: Implemented the suggested changes.

**Pull request merged**